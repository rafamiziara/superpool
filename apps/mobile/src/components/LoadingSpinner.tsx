import React from 'react';
import { ActivityIndicator, View } from 'react-native';

interface LoadingSpinnerProps {
  size?: 'small' | 'large';
  className?: string;
  color?: string;
}

export function LoadingSpinner({ 
  size = 'large', 
  className = '', 
  color = '#2563eb' // Default to primary color
}: LoadingSpinnerProps) {
  return (
    <View className={`items-center justify-center ${className}`}>
      <ActivityIndicator size={size} color={color} />
    </View>
  );
}