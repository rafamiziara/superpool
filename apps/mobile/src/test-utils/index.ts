/**
 * Centralized export of all test utilities
 * Import from this file to get all testing helpers
 */

export * from './mockStores'
export * from './testProviders'
export * from './renderWithStore'

/**
 * Common test utilities and helpers
 */

/**
 * Wait for MobX reactions to settle
 * Useful when testing computed values or reactions
 */
export const waitForMobX = async (timeout = 100): Promise<void> => {
  return new Promise((resolve) => {
    setTimeout(resolve, timeout)
  })
}

/**
 * Mock Firebase Auth user object for testing
 */
export const createMockFirebaseUser = (overrides: Record<string, unknown> = {}) => ({
  uid: 'test-user-id',
  email: 'test@example.com',
  displayName: 'Test User',
  emailVerified: true,
  photoURL: null,
  phoneNumber: null,
  ...overrides,
})

/**
 * Mock Ethereum address for testing
 */
export const MOCK_ETH_ADDRESS = '0x1234567890123456789012345678901234567890'
export const MOCK_ETH_ADDRESS_2 = '0x0987654321098765432109876543210987654321'

/**
 * Common test chain IDs
 */
export const TEST_CHAIN_IDS = {
  ETHEREUM_MAINNET: 1,
  POLYGON_MAINNET: 137,
  POLYGON_AMOY: 80002,
  LOCALHOST: 31337,
} as const
