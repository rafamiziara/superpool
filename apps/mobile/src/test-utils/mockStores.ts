import { AuthenticationStore } from '../stores/AuthenticationStore'
import { WalletStore } from '../stores/WalletStore'
import { PoolManagementStore } from '../stores/PoolManagementStore'
import { UIStore } from '../stores/UIStore'
import { RootStore } from '../stores/RootStore'
import { ErrorType } from '../utils/errorHandling'

/**
 * Factory functions for creating mock MobX stores with preset states
 */

export const createMockAuthenticationStore = (overrides: Partial<AuthenticationStore> = {}): AuthenticationStore => {
  const store = new AuthenticationStore()
  
  // Reset the store to clean state first
  store.completedSteps.clear()
  store.completedSteps.add('connect-wallet') // resetProgress adds this
  
  // Create a mocked version by wrapping the original methods
  const originalStartStep = store.startStep.bind(store)
  const originalCompleteStep = store.completeStep.bind(store) 
  const originalFailStep = store.failStep.bind(store)
  const originalResetProgress = store.resetProgress.bind(store)
  const originalGetStepStatus = store.getStepStatus.bind(store)
  const originalGetStepInfo = store.getStepInfo.bind(store)
  const originalGetAllSteps = store.getAllSteps.bind(store)

  // Create spies that wrap the original functionality
  const mockStartStep = jest.fn().mockImplementation(originalStartStep)
  const mockCompleteStep = jest.fn().mockImplementation(originalCompleteStep)
  const mockFailStep = jest.fn().mockImplementation(originalFailStep)
  const mockResetProgress = jest.fn().mockImplementation(originalResetProgress)
  const mockGetStepStatus = jest.fn().mockImplementation(originalGetStepStatus)
  const mockGetStepInfo = jest.fn().mockImplementation(originalGetStepInfo)
  const mockGetAllSteps = jest.fn().mockImplementation(originalGetAllSteps)

  // Replace methods on the store with mocks
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  ;(store as any).startStep = mockStartStep
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  ;(store as any).completeStep = mockCompleteStep
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  ;(store as any).failStep = mockFailStep
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  ;(store as any).resetProgress = mockResetProgress
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  ;(store as any).getStepStatus = mockGetStepStatus
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  ;(store as any).getStepInfo = mockGetStepInfo
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  ;(store as any).getAllSteps = mockGetAllSteps

  // Apply overrides
  Object.assign(store, overrides)

  return store
}

export const createMockWalletStore = (overrides: Partial<WalletStore> = {}): WalletStore => {
  const store = new WalletStore()

  // Apply overrides
  Object.assign(store, overrides)

  return store
}

export const createMockPoolManagementStore = (overrides: Partial<PoolManagementStore> = {}): PoolManagementStore => {
  const store = new PoolManagementStore()

  // Apply overrides
  Object.assign(store, overrides)

  return store
}

export const createMockUIStore = (overrides: Partial<UIStore> = {}): UIStore => {
  const store = new UIStore()

  // Apply overrides
  Object.assign(store, overrides)

  return store
}

export const createMockRootStore = (
  storeOverrides: {
    authenticationStore?: Partial<AuthenticationStore>
    walletStore?: Partial<WalletStore>
    poolManagementStore?: Partial<PoolManagementStore>
    uiStore?: Partial<UIStore>
  } = {}
): RootStore => {
  const rootStore = new RootStore()

  // Always replace authenticationStore with mocked version for tests
  rootStore.authenticationStore = createMockAuthenticationStore(storeOverrides.authenticationStore || {})

  if (storeOverrides.walletStore) {
    rootStore.walletStore = createMockWalletStore(storeOverrides.walletStore)
  }

  if (storeOverrides.poolManagementStore) {
    rootStore.poolManagementStore = createMockPoolManagementStore(storeOverrides.poolManagementStore)
  }

  if (storeOverrides.uiStore) {
    rootStore.uiStore = createMockUIStore(storeOverrides.uiStore)
  }

  return rootStore
}

/**
 * Preset mock store configurations for common test scenarios
 */

export const mockStorePresets = {
  // Authenticated user with connected wallet
  authenticatedWithWallet: () =>
    createMockRootStore({
      walletStore: {
        isConnected: true,
        address: '0x1234567890123456789012345678901234567890',
        chainId: 137,
        connectionError: null,
        isConnecting: false,
      },
    }),

  // Unauthenticated user
  unauthenticated: () =>
    createMockRootStore({
      walletStore: {
        isConnected: false,
        address: undefined,
        chainId: undefined,
        connectionError: null,
        isConnecting: false,
      },
    }),

  // Authentication in progress
  authenticating: () =>
    createMockRootStore({
      authenticationStore: {
        authLock: {
          isLocked: true,
          startTime: Date.now(),
          walletAddress: '0x1234567890123456789012345678901234567890',
          abortController: new AbortController(),
          requestId: 'test-request-id',
        },
      },
    }),

  // Wallet connection in progress
  connectingWallet: () =>
    createMockRootStore({
      walletStore: {
        isConnected: false,
        address: undefined,
        chainId: undefined,
        connectionError: null,
        isConnecting: true,
      },
    }),

  // Error states
  authenticationError: () =>
    createMockRootStore({
      authenticationStore: {
        authError: {
          name: 'AppError',
          message: 'Authentication failed',
          type: ErrorType.AUTHENTICATION_FAILED,
          userFriendlyMessage: 'Authentication failed',
        },
      },
    }),

  walletConnectionError: () =>
    createMockRootStore({
      walletStore: {
        isConnected: false,
        address: undefined,
        chainId: undefined,
        connectionError: 'Failed to connect wallet',
        isConnecting: false,
      },
    }),
}
