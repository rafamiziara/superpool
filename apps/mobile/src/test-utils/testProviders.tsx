import React, { ReactNode } from 'react'
import { StoreContext } from '../stores/StoreContext'
import { RootStore } from '../stores/RootStore'
import { createMockRootStore } from './mockStores'

/**
 * Test provider that wraps components with a mock MobX store context
 */
interface TestStoreProviderProps {
  children: ReactNode
  store?: RootStore
}

export const TestStoreProvider: React.FC<TestStoreProviderProps> = ({ children, store = createMockRootStore() }) => {
  // Create StoreContext.Provider manually since we need to pass our mock store
  const StoreContextProvider = StoreContext.Provider as React.ComponentType<{
    value: RootStore
    children: ReactNode
  }>

  return <StoreContextProvider value={store}>{children}</StoreContextProvider>
}

/**
 * Higher-order component that wraps a component with TestStoreProvider
 */
export const withMockStore = <P extends object>(Component: React.ComponentType<P>, store?: RootStore) => {
  const WrappedComponent: React.FC<P> = (props) => (
    <TestStoreProvider store={store}>
      <Component {...props} />
    </TestStoreProvider>
  )

  WrappedComponent.displayName = `withMockStore(${Component.displayName || Component.name})`

  return WrappedComponent
}
