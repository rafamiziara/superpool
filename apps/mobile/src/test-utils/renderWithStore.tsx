import { render, renderHook, RenderOptions, RenderResult } from '@testing-library/react-native'
import React, { ReactElement } from 'react'
import { RootStore } from '../stores/RootStore'
import { createMockRootStore } from './mockStores'
import { TestStoreProvider } from './testProviders'

/**
 * Custom render options that includes store configuration
 */
interface CustomRenderOptions extends Omit<RenderOptions, 'wrapper'> {
  store?: RootStore
  wrapper?: React.ComponentType<{ children: React.ReactNode }>
}

/**
 * Custom render function that wraps components with TestStoreProvider
 * This is the recommended way to render components that use MobX stores in tests
 */
export const renderWithStore = (ui: ReactElement, options: CustomRenderOptions = {}): RenderResult => {
  const { store = createMockRootStore(), wrapper: CustomWrapper, ...renderOptions } = options

  const Wrapper: React.FC<{ children: React.ReactNode }> = ({ children }) => {
    if (CustomWrapper) {
      return (
        <TestStoreProvider store={store}>
          <CustomWrapper>{children}</CustomWrapper>
        </TestStoreProvider>
      )
    }

    return <TestStoreProvider store={store}>{children}</TestStoreProvider>
  }

  return render(ui, { wrapper: Wrapper, ...renderOptions })
}

/**
 * Re-export everything from React Native Testing Library for convenience
 */
export * from '@testing-library/react-native'

/**
 * Override the default render method to always use renderWithStore
 * This ensures all components are rendered with a mock store context
 */
export { renderWithStore as render }

/**
 * Custom renderHook function that wraps hooks with TestStoreProvider
 * This is the recommended way to test hooks that use MobX stores
 */
export const renderHookWithStore = <TResult,>(hook: () => TResult, options: { store?: RootStore } = {}) => {
  const { store = createMockRootStore() } = options

  const Wrapper: React.FC<{ children: React.ReactNode }> = ({ children }) => {
    return <TestStoreProvider store={store}>{children}</TestStoreProvider>
  }

  return renderHook(() => hook(), { wrapper: Wrapper })
}

/**
 * Utility function to create a minimal render for non-store components
 * Use this only when you need to test components that don't use MobX stores
 */
export const renderWithoutStore = render as typeof import('@testing-library/react-native').render
