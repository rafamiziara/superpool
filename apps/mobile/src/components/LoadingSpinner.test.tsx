import React from 'react'
import { render } from '../test-utils'
import { LoadingSpinner } from './LoadingSpinner'

describe('LoadingSpinner', () => {
  describe('Default Props', () => {
    it('should render with default props', () => {
      const { getByTestId } = render(<LoadingSpinner />)

      const activityIndicator = getByTestId('activity-indicator')
      expect(activityIndicator.props.size).toBe('large')
      expect(activityIndicator.props.color).toBe('#2563eb')
    })

    it('should render with default className', () => {
      const { getByTestId } = render(<LoadingSpinner />)

      const container = getByTestId('loading-spinner-container')
      expect(container.props.className).toContain('items-center justify-center')
    })
  })

  describe('Custom Props', () => {
    it('should render with small size', () => {
      const { getByTestId } = render(<LoadingSpinner size="small" />)

      const activityIndicator = getByTestId('activity-indicator')
      expect(activityIndicator.props.size).toBe('small')
    })

    it('should render with large size', () => {
      const { getByTestId } = render(<LoadingSpinner size="large" />)

      const activityIndicator = getByTestId('activity-indicator')
      expect(activityIndicator.props.size).toBe('large')
    })

    it('should render with custom color', () => {
      const customColor = '#ff0000'
      const { getByTestId } = render(<LoadingSpinner color={customColor} />)

      const activityIndicator = getByTestId('activity-indicator')
      expect(activityIndicator.props.color).toBe(customColor)
    })

    it('should render with custom className', () => {
      const customClassName = 'my-custom-class p-4'
      const { getByTestId } = render(<LoadingSpinner className={customClassName} />)

      const container = getByTestId('loading-spinner-container')
      expect(container.props.className).toContain('items-center justify-center')
      expect(container.props.className).toContain(customClassName)
    })

    it('should combine default and custom className', () => {
      const customClassName = 'p-4 bg-gray-100'
      const { getByTestId } = render(<LoadingSpinner className={customClassName} />)

      const container = getByTestId('loading-spinner-container')
      const className = container.props.className

      expect(className).toContain('items-center justify-center')
      expect(className).toContain('p-4')
      expect(className).toContain('bg-gray-100')
    })
  })

  describe('Accessibility', () => {
    it('should have proper accessibility properties', () => {
      const { getByTestId } = render(<LoadingSpinner />)

      const activityIndicator = getByTestId('activity-indicator')
      expect(activityIndicator).toBeTruthy()
    })

    it('should be accessible with screen readers', () => {
      const { getByTestId } = render(<LoadingSpinner />)

      const container = getByTestId('loading-spinner-container')
      expect(container).toBeTruthy()

      const activityIndicator = getByTestId('activity-indicator')
      expect(activityIndicator).toBeTruthy()
    })
  })

  describe('Snapshot Testing', () => {
    it('should match snapshot with default props', () => {
      const component = render(<LoadingSpinner />)
      expect(component.toJSON()).toMatchSnapshot()
    })

    it('should match snapshot with custom props', () => {
      const component = render(<LoadingSpinner size="small" color="#00ff00" className="p-2 m-4" />)
      expect(component.toJSON()).toMatchSnapshot()
    })
  })

  describe('Props Validation', () => {
    it('should handle all size variants correctly', () => {
      const sizes: Array<'small' | 'large'> = ['small', 'large']

      sizes.forEach((size) => {
        const { getByTestId } = render(<LoadingSpinner size={size} />)
        const activityIndicator = getByTestId('activity-indicator')
        expect(activityIndicator.props.size).toBe(size)
      })
    })

    it('should handle empty className gracefully', () => {
      const { getByTestId } = render(<LoadingSpinner className="" />)

      const container = getByTestId('loading-spinner-container')
      expect(container.props.className).toBe('items-center justify-center ')
    })

    it('should handle undefined props gracefully', () => {
      const { getByTestId } = render(<LoadingSpinner size={undefined} color={undefined} />)

      const activityIndicator = getByTestId('activity-indicator')
      expect(activityIndicator.props.size).toBe('large') // Default fallback
      expect(activityIndicator.props.color).toBe('#2563eb') // Default fallback
    })
  })
})
