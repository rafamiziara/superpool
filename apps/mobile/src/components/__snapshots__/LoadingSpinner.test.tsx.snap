// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`LoadingSpinner Snapshot Testing should match snapshot with custom props 1`] = `
<View
  className="items-center justify-center p-2 m-4"
  testID="loading-spinner-container"
>
  <ActivityIndicator
    color="#00ff00"
    size="small"
    testID="activity-indicator"
  />
</View>
`;

exports[`LoadingSpinner Snapshot Testing should match snapshot with default props 1`] = `
<View
  className="items-center justify-center "
  testID="loading-spinner-container"
>
  <ActivityIndicator
    color="#2563eb"
    size="large"
    testID="activity-indicator"
  />
</View>
`;
